### PR TITLE
SortModifiers rewrite fixes + doc prettyfying

### LINKED ISSUE #1148: New SortModifiers overwrites modifiers of implicit class parameter val/vars
- **Version**: 1.5.0
- **Integration**: IntelliJ, SBT
- **Configuration**:
```
rewrite {
  rules = [SortModifiers]
  sortModifiers {
    #the crucial part of this bug is for modifiers to have higher precedence than "implicit"
    order = ["private", "protected", "implicit", "final", "sealed", "abstract", "override", "lazy"]
  }
}
```

## Steps

Given code like this:
```scala
class Test(private val x: Int)(
  implicit
  private val i1: Int,
  private val i2: String
)
```

When I run scalafmt anywhere (IntelliJ, or SBT).

## Problem

Scalafmt formats code like this:
```scala
class Test2(private val x: Int)(
  implicit
  implicit val i1: Int,
  implicit val i2: String
)
```

## Expectation

I would like the formatted output to look like this:
```scala
class Test2(private val x: Int)(
  implicit
  private val i1: Int,
  private val i2: String
)
```

## Workaround

I've found that this happens only if the class parameters have modifiers that have higher precedence than `implicit`.

### PR DESCRIPTION
**Pull request by lorandszakacs**
Closes #1148 (bug), #1149 (docs)

There is one super weird corner case that is not handled properly by the fix. There's also a test for it:
If you have a sort order where `implicit` has lower order than the rest, then the following code gets formatted weirdly.

Config:
```
rewrite {
  rules = [SortModifiers]
  sortModifiers {
    order = ["private", "protected" , "abstract", "final", "sealed", "implicit", "override", "lazy"]
  }
}
```
Original code:
```
class X(private final val x: Int)(
    implicit final private val i1: Int,
    implicit final private var i2: Int
)
```
Formatted code:
```
class X(private final val x: Int)(
    implicit private final val i1: Int,
    private final implicit var i2: Int
)
```

But this is a super weird corner case where the `implicit` keyword is duplicated, and honestly until I wrote this format I had no idea that was even possible 😅. 

You will also notice that this implementation is actually cleaner than the buggy one, which is a pleasant surprise 😃 
/cc @olafurpg

### CODE REVIEW

File: scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/SortModifiers.scala [hunk 1]
```diff
@@ -35,31 +35,48 @@ object SortModifiers extends Rewrite {
...
+    *     private[this] var i2: String
+    * )
+    * }}}
+    *
+    * `val i1`, and `var i2` have a ``Mod.Implicit`` with empty tokens.
+    * Therefore we want to completely ignore this "mod" whenever we sort,
+    * and apply patches
+    */
+  private def isHiddenImplicit(m: Mod): Boolean = {
+    m.tokens.isEmpty && m.is[Mod.Implicit]
```
Comment (olafurpg) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/SortModifiers.scala]: I think this will return true for `def foo(implicit x: Int)` due to a bug, can you double check?

Comment (lorandszakacs) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/SortModifiers.scala]: @olafurpg will do.

Comment (lorandszakacs) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/SortModifiers.scala]: @olafurpg indeed it does return `true`, but luckily no rewrite is necessary in this case, so the final result is OK. Added an explicit test case for this:
```
<<< 1148 — only implicit param
def foo(implicit x: Int)
>>>
def foo(implicit x: Int)
```

**Comment by olafurpg** [comment]
This came out nice 👍 

screen shot 2018-04-24 at 13 37 02

**Comment by olafurpg** [comment]
I'll try to get a hotfix out later this week, I want to address #1147 as well
